File: OpenFlix-iOS/OpenFlix-iOS/Repositories/AuthRepository.swift
import Foundation

@MainActor
class AuthRepository: ObservableObject {
    private let api = OpenFlixAPI.shared
    private let keychain = KeychainHelper.shared

    @Published var currentUser: User?
    @Published var currentProfile: Profile?
    @Published var profiles: [Profile] = []

    // MARK: - Authentication

    func login(username: String, password: String) async throws -> User {
        let response = try await api.login(username: username, password: password)

        // Save token
        keychain.saveToken(response.token)
        await api.setToken(response.token)

        let user = response.user.toDomain()
        currentUser = user
        return user
    }

    func register(name: String, email: String, password: String) async throws -> User {
        let response = try await api.register(name: name, email: email, password: password)

        // Save token
        keychain.saveToken(response.token)
        await api.setToken(response.token)

        let user = response.user.toDomain()
        currentUser = user
        return user
    }

    func logout() async {
        try? await api.logout()
        keychain.deleteToken()
        await api.setToken(nil)
        currentUser = nil
        currentProfile = nil
        profiles = []
        UserDefaults.standard.currentProfileUUID = nil
    }

    func checkAuthStatus() async -> Bool {
        guard keychain.getToken() != nil else { return false }

        do {
            let user = try await api.getUser()
            currentUser = user.toDomain()
            return true
        } catch {
            // Token invalid, clear it
            keychain.deleteToken()
            await api.setToken(nil)
            return false
        }
    }

    // MARK: - Profiles

    func loadProfiles() async throws {
        let response = try await api.getHomeUsers()
        profiles = response.users.map { $0.toDomain() }

        // Try to restore last selected profile
        if let savedUUID = UserDefaults.standard.currentProfileUUID,
           let profile = profiles.first(where: { $0.uuid == savedUUID }) {
            currentProfile = profile
        }
    }

    func switchProfile(_ profile: Profile, pin: String? = nil) async throws {
        let response = try await api.switchProfile(uuid: profile.uuid, pin: pin)

        // Update token if provided (Android uses authToken, check both)
        if let newToken = response.authToken ?? response.token {
            keychain.saveToken(newToken)
            await api.setToken(newToken)
        }

        currentProfile = profile
        UserDefaults.standard.currentProfileUUID = profile.uuid
    }

    func createProfile(name: String, isKid: Bool, pin: String?) async throws {
        // Note: This requires admin permissions
        // Implementation depends on profile creation endpoint response
    }
}

File: openflix-kmp/shared/src/commonMain/kotlin/com/openflix/data/repository/AuthRepository.kt
package com.openflix.data.repository

import com.openflix.data.dto.AuthResponse
import com.openflix.data.network.OpenFlixApi
import com.openflix.domain.model.Profile
import com.openflix.domain.model.User
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class AuthRepository(private val api: OpenFlixApi) {
    private val _currentUser = MutableStateFlow<User?>(null)
    val currentUser: StateFlow<User?> = _currentUser.asStateFlow()

    private val _currentProfile = MutableStateFlow<Profile?>(null)
    val currentProfile: StateFlow<Profile?> = _currentProfile.asStateFlow()

    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()

    private var token: String? = null

    suspend fun login(username: String, password: String): AuthResponse {
        val response = api.login(username, password)
        token = response.token
        api.setToken(response.token)
        _currentUser.value = response.user.toDomain()
        _isAuthenticated.value = true
        return response
    }

    suspend fun logout() {
        try { api.logout() } catch (_: Exception) {}
        token = null
        api.setToken(null)
        _currentUser.value = null
        _currentProfile.value = null
        _isAuthenticated.value = false
    }

    suspend fun getUser(): User {
        val dto = api.getUser()
        val user = dto.toDomain()
        _currentUser.value = user
        return user
    }

    suspend fun getProfiles(): List<Profile> {
        return api.getProfiles().map { it.toDomain() }
    }

    suspend fun switchProfile(uuid: String, pin: String? = null): Boolean {
        val response = api.switchProfile(uuid, pin)
        val newToken = response.authToken ?: response.token
        if (newToken != null) {
            token = newToken
            api.setToken(newToken)
        }
        return response.success ?: (newToken != null)
    }

    fun restoreSession(serverUrl: String, savedToken: String) {
        api.configure(serverUrl, savedToken)
        token = savedToken
        _isAuthenticated.value = true
    }
}
